File: python-task/fixture_test.py
"""Test that loads expected messages from a fixture file using runfiles."""
import json
import unittest

from python.runfiles import runfiles

from subtask_a.subtask_a import get_message as get_a
from subtask_b.subtask_b import get_message as get_b


class TestFixture(unittest.TestCase):
    def setUp(self):
        r = runfiles.Create()
        path = r.Rlocation("xp_bazel/python-task/testdata/expected_messages.json")
        with open(path) as f:
            self.expected = json.load(f)

    def test_subtask_a_matches_fixture(self):
        self.assertEqual(get_a(), self.expected["subtask_a"])

    def test_subtask_b_matches_fixture(self):
        self.assertEqual(get_b(), self.expected["subtask_b"])


if __name__ == "__main__":
    unittest.main()

File: go-task/fixture_test.go
package main

import (
	"encoding/json"
	"os"
	"testing"

	"github.com/bazelbuild/rules_go/go/runfiles"
	subtaska "xp_bazel/go-task/subtask-a"
	subtaskb "xp_bazel/go-task/subtask-b"
)

func TestFixture(t *testing.T) {
	path, err := runfiles.Rlocation("xp_bazel/go-task/testdata/expected_messages.json")
	if err != nil {
		t.Fatalf("failed to find runfile: %v", err)
	}

	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("failed to read fixture: %v", err)
	}

	var expected map[string]string
	if err := json.Unmarshal(data, &expected); err != nil {
		t.Fatalf("failed to parse fixture: %v", err)
	}

	if got := subtaska.GetMessage(); got != expected["subtask_a"] {
		t.Errorf("subtask_a: expected %q, got %q", expected["subtask_a"], got)
	}

	if got := subtaskb.GetMessage(); got != expected["subtask_b"] {
		t.Errorf("subtask_b: expected %q, got %q", expected["subtask_b"], got)
	}
}
